File: Java/1609. Even Odd Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isEvenOddTree(TreeNode root) {
        int level = 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        
        while (!q.isEmpty()) {
            if (level % 2 == 0) {
                if (!checkEven(q)) {
                    return false;
                }
            } else {
                if (!checkOdd(q)) {
                    return false;
                }
            }
            level++;
        }
        return true;
    }

    private boolean checkOdd(Queue<TreeNode> q) {
        int size = q.size(), pre = Integer.MAX_VALUE;

        // odd level, nums must be even and decreasing
        for (int i = 0; i < size; i++) {
            TreeNode node = q.poll();
            if (node.val % 2 == 1 || node.val >= pre) {
                return false;
            }
            pre = node.val;
            if (node.left != null) {
                q.offer(node.left);
            }
            if (node.right != null) {
                q.offer(node.right);
            }
        }
        return true;
    }

    private boolean checkEven(Queue<TreeNode> q) {
        int size = q.size(), pre = 0;

        // odd level, nums must be even and decreasing
        for (int i = 0; i < size; i++) {
            TreeNode node = q.poll();
            if (node.val % 2 == 0 || node.val <= pre) {
                return false;
            }
            pre = node.val;
            if (node.left != null) {
                q.offer(node.left);
            }
            if (node.right != null) {
                q.offer(node.right);
            }
        }
        return true;
    }
}

File: Python3/1609. Even Odd Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        
        def checkOdd(queue) -> bool:
            size = len(queue)
            pre = 10 ** 6 + 1
            for _ in range(size):
                node = queue.popleft()
                # odd level num must be even and decreasing
                if node.val % 2 or node.val >= pre:
                    return False
                pre = node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            return True

        def checkEven(queue) -> bool:
            size = len(queue)
            pre = 0
            for _ in range(size):
                node = queue.popleft()
                # even level num must be odd and increasing
                if node.val % 2 == 0 or node.val <= pre:
                    return False
                pre = node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            return True

        level = 0
        q = deque([root])

        while q:
            # odd level
            if level % 2:
                if not checkOdd(q):
                    return False
            else:
                if not checkEven(q):
                    return False
            level += 1

        return True